src/main/scala/essential_scala/Ex4Traits.scala

```diff
@@ -56,4 +58,100 @@ object Ex4Traits extends App {
   }
 
   //  4.5.6.3 Email
+  trait Visitor {
+    def id: String
+    def createdAt: Date
+    def age: Long = new Date().getTime - createdAt.getTime
+  }
+  case class Anonymous(
+    id: String,
+    createdAt: Date = new Date()
+  ) extends Visitor
+  case class User(
+    id: String,
+    email: String,
+    createdAt: Date = new Date()
+  ) extends Visitor
+  // not only depends on other fields and methods in a class
+  //  we generally prefer sealed traits
+  object EmailService {
+    def email(visitor: Visitor): Unit = {
+      visitor match {
+        case a: Anonymous => ()
+        case u: User      => println("send " + u.email)
+      }
+    }
+  }
+
+  // 4.6.3.2 The Forest of Trees
+  // pattern matching
+  sealed trait Tree {}
+  case class Node(left: Tree, right: Tree) extends Tree
+  case class Leaf(element: Int) extends Tree
+  object Math {
+    def prettyPrint(t: Tree, indent: String): Unit = t match {
```

Comment:
Why `indent` is not a number? Usually it's just the number of spaces.

Reply:
And chapter 4.6.2 is about tail recursion. This is not tail recursion

Reply:
Fixed